### ISSUE TITLE
Lesson 18, the for-loop is NOT doing the same as the while-loop

### ISSUE THREAD

**Issue by sgoman** [GDQuest/learn-gdscript#681]
The while-loop checks the current position each time, so it works for all cells on the board. It only loops the required amount to reach the right edge.
The for-loop always walks board_size.x times to the right and therefore only works for starting positions on the left edge of the board. If the starting cell is at least one column away from the left edge, it over shoots.

It *could* be remedied by changing the loop to this:

```
for number in range(board_size.x - cell.x - 1):
    cell.x += 1
```

But I doubt you could still call that one a little simpler than the while-loop variant.

**Comment by NathanLovato** [comment]
The character always starts on the left of the board, so both examples do the same thing in practice. That it wouldn't work the same in another context is beyond the scope of what we're trying to convey.

Please keep in mind that this course is for programming neophytes. What we care about is that students understand the lessons and can complete the practices. We cheat with many things for learning purposes, as you might've noticed.

If you have an idea for an alternate two-line example that'd still be visual and very game-like (like moving in a board game) don't hesitate to share!

If we don't have a better alternative, I think the example is okay as it is.

**Comment by sgoman** [comment]
Yeah, I get what you mean. It's the concept of "lying to children" to hide complexity beyond their current ability. Some kind of "need to know" limitation on the information provided. Every IT topic struggles with various layers of abstraction and which layer to settle on for the discussion at hand. And aiming for perfection means getting rid of anything that is not strictly necessary.

Still, I would prefer to soften the expression of them being identical, maybe as simple as adding "in this scenario" or a little explanation like "as our character always starts on the opposite side of the board". I interpreted the explanation on a more general level, your distinction of the limited scope was not obvious to me.

Another part where I was raising eyebrows was the statement that for-loops could not result in infinite loops. All three parts of the loop statement (initialization, condition and action) are optional and in addition to that you could also tweak the element the condition works on in the loop body (bad coding habit, sure, but nothing prevents you from doing it).
A newcomer might simply take away that for-loops are superior, so why should they bother learning about the inferior while-loop at all? Again, I agree that context matters to make the choice of which loop construct to chose.

Maybe just room for improvement, nice to have. Polish for later stages of development.

**Comment by NathanLovato** [comment]
I see, makes sense, thanks!

**Comment by ron0studios** [comment]
I made some small edits, if there's anything else @sgoman , please let me know thanks!
